### PR TITLE
Add postcondition tracking to transactions

### PR DESCRIPTION
**Pull request by wilhuff**
This makes it so that certain sequences of operations now work. Namely:

  * Read non-existent document, then set, set
  * Read existing document, delete, set

### CODE REVIEW

File: Firestore/core/src/firebase/firestore/core/transaction.mm [hunk 1]
```diff
@@ -159,6 +189,7 @@
   // Since the delete will be applied before all following writes, we need to
   // ensure that the precondition for the next write will be exists: false.
   read_versions_[key] = SnapshotVersion::None();
```
Comment (fredzqm) [inline Firestore/core/src/firebase/firestore/core/transaction.mm]: It looks like updating `read_versions_` is no longer necessary.

File: Firestore/Example/Tests/Integration/FSTTransactionTests.mm [hunk 2]
```diff
@@ -163,24 +163,70 @@ - (void)testDeleteDocumentAndSet {
...
+        XCTAssertEqualObjects(@YES, result);
+        XCTAssertNil(error);
+        [expectation fulfill];
+      }];
+  [self awaitExpectations];
+
+  FIRDocumentSnapshot *snapshot = [self readDocumentForRef:doc];
+  XCTAssertEqualObjects(snapshot.data, @{@"c" : @"d"});
+}
+
```
[the remark this answers is not in the document]
Comment (fredzqm) [reply Firestore/Example/Tests/Integration/FSTTransactionTests.mm]: Yeah, there could also be more cases involved considering the _read_versions = 0 or not.
3 (read_version: no read, exists, missing) * 3 (update, set, delete) * 3 (update, set, delete)

Table tests would be great.

**Comment by fredzqm** [comment]
FWIW, the simplest solution is "only append preconditions in the first write, and let the backend do the validations".

I don't think there is much value in client-side validations because these are 400s (i.e. they wrote a bug).

**Comment by fredzqm** [comment]
If verify w/ mutation collapse is launched, it can be even simpler:
prepend all the reads as verify, and the backend should perform mutation collapse for you.

IMO, it may not be worthwhile to reimplement the mutation collapse validation in the client side as well.

**Comment by wilhuff** [comment]
Based on offline discussion, we're not going to proceed with this approach. A simpler one will work better.

**Pull request closed without merging**